Rework CSI driver installation to CVO / cluster-storage-operator

jsafrane opened the pull request:
Since installation of CSI drivers shows to be challenging, this PR moves CSI driver installation to CVO / cluster-storage-operator.

Upgrade from 4.5 (with AWS EBS CSI driver installed through OLM) to 4.6 (where it's adopted by cluster-storage-operator) is explicitly covered.

Installation through OLM is mentioned as "Alternative considered", with the old enhancement there.

@openshift/storage, @openshift/openshift-architects PTAL

jsafrane commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -79,61 +79,187 @@ along with in-tree driver and users could use both.
+  and [`OperatorStatus`](https://github.com/openshift/api/blob/95abe2d2f4223d5931e418bf8e4d3773d16b42c0/operator/v1/types.go#L97).
+  In the future, each CSI driver may introduce its own configuration fields.
+* Each CSI driver runs in a dedicated namespace, typically `openshift-<backed>-csi-driver` (`openshift-aws-ebs-csi-driver`).
+  * TODO: or run everything in cluster-storage namespace? OLM installs into openshift-aws-ebs-csi-driver, we would need to delete it before starting a new driver in cluster-storage. 
+* TODO: where the operator runs? cluster-storage namespace?
```
There are two open items for @openshift/storage team about namespaces where the operator / CSI driver will run.

jsafrane commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -79,61 +79,187 @@ along with in-tree driver and users could use both.
+  and [`OperatorStatus`](https://github.com/openshift/api/blob/95abe2d2f4223d5931e418bf8e4d3773d16b42c0/operator/v1/types.go#L97).
+  In the future, each CSI driver may introduce its own configuration fields.
+* Each CSI driver runs in a dedicated namespace, typically `openshift-<backed>-csi-driver` (`openshift-aws-ebs-csi-driver`).
+  * TODO: or run everything in cluster-storage namespace? OLM installs into openshift-aws-ebs-csi-driver, we would need to delete it before starting a new driver in cluster-storage. 
```
Currently all  CSI drivers are the same. Meaning AWS EBS and Manila. And it looks like the others will not have any special permissions too, however, I don't want to paint us in the corner.

jwforres commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -79,61 +79,187 @@ along with in-tree driver and users could use both.
+2. 4.6 CSO detects that there is AWS / Manila driver installed by OLM.
+3. CSO deletes Subscription of the operator. OLM removes the operator, but leaves the CRD, CR and the driver / operand running.
+4. CSO runs the new CSI driver operator.
+5. The CSI driver operator adopts the old operand - it reconciles the old CSI driver objects (Deployment, DaemonSet,
```
As far as I understand these operators are going out as tech previews. We do not need to make the tech previews upgrade to the GA. Instead of adopting and upgrading the tech preview functionality, we should consider documenting that the administrator needs to uninstall the tech preview before upgrading to 4.6. A few things to consider though if we went down that route:

1. Uninstalling the OLM operator (deleting the Subscription) won't cascade to uninstall the operand or anything else it was managing, in this case things like the StorageClass. What are the potential risks if things like the StorageClass and running drivers aren't cleaned up correctly and the CSO installed operator isn't able to adopt them?

2. What are the implications of asking the administrator to completely uninstall the drivers before upgrading? Will that result in losing data in PVs that had been backed by that driver? Do the PVs get deleted? Would we be orphaning the EBS volumes in some way?

Another tool we have are upgrade preflight checks. If we needed to we could write a preflight that makes sure the operator, APIs, etc don't exist on the cluster and prevent the upgrade until the administrator removes it. 

We will also be able to monitor how many installations of these tech preview operators exist through telemetry once they are shipped, this would help us weigh the decision of whether it is worth implementing a preflight, or doing the adoption of the tech preview.

jsafrane replied:
Unfortunately, Manila wants to be GA in 4.5. We already missed 4.4 where they wanted to be TP.
cc @Fedosin

jsafrane commented:
* Added note about Manila GA / AWS EBS TP in 4.5.
* Added note about un-installation (basically, you can't remove a CSI driver managed by cluster-storage-operator easily).

jsafrane commented:
Added RHV / oVirt CSI driver to 4.6.

bertinatto commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -196,12 +196,65 @@ Kubernetes 1.21 is current upstream target for removal of in-tree cloud provider
+In this case, it's important that the CSI driver's StorageClass deployed by CSO has the same name as the one deployed by the CSI driver operator.
+
+This alternative migh impose another problem: AWS EBS CSI Driver Operator doesn't remove the operand if there are volumes using the CSI driver.
+
```
@openshift/storage I pushed commit 477b905d to capture what we discussed yesterday. PTAL

jsafrane commented:
Updated after storage team design discussions. Main changes:

* Move all CSI driver operators + the drivers into `cluster-storage-operator` namespace.
  * We expect 1-2 CSI drivers per cloud.
  * Allows us to have more restrictive RBAC rules.
  * Allows us to have single namespace for all storage stuff in OCP.
  * Complicates migration from OLM.

(Moving to individual namespaces for each driver is a viable option)

* Strictly specified what CSI drivers this ehancement covers (cloud-based ones except IBM, which can be added later).

bertinatto commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -116,21 +147,28 @@ in openshift-aws-ebs-csi-driver namespace / openshift-manila-csi-driver.
-   RBAC, ...) with updated 4.6 content. I.e., all objects created by the 4.6 operator must have the same name
-   as the objects created by the old 4.5 operator.
+4. CSO deletes the old namespace (`openshift-aws-ebs-csi-driver` / `openshift-manila-csi-driver`).
+5. CSO runs the new CSI driver operator in `cluster-storage-operatator` namespace.
```
typo: operatator

jsafrane replied:
Fixed

bertinatto commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -116,21 +147,28 @@ in openshift-aws-ebs-csi-driver namespace / openshift-manila-csi-driver.
-5. The CSI driver operator adopts the old operand - it reconciles the old CSI driver objects (Deployment, DaemonSet,
-   RBAC, ...) with updated 4.6 content. I.e., all objects created by the 4.6 operator must have the same name
-   as the objects created by the old 4.5 operator.
+4. CSO deletes the old namespace (`openshift-aws-ebs-csi-driver` / `openshift-manila-csi-driver`).
```
What about other cluster-scoped objects, like RBAC rules and cloud credentials (they have a reference to the operand namespace)? Should we delete them or just expect the new operator to resync and change the namespaces they reference to?

IMO it's better to follow the old namespace approach, i.e., delete them.

 created in `openshift-cloud-credential-operator` namespace?

jsafrane replied:
Added note about cluster-scoped object. Delete everything except for StorageClass and CR.

bertinatto commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -144,7 +182,8 @@ Bases on current upstream plans & assumptions.
 * We support installation of AWS EBS (tech preview) and Manila CSI (GA) drivers with in-tree drivers. User can use both
-  drivers at the same time. The default storage class is for in-tree volume plugin (unless cluster admin changes it).
+  in-tree volume plugin and CSI drivers at the same time. The default storage class is for in-tree volume plugin (unless
+  cluster admin changes it).
```
nit: there's no in-tree plugin for Manila

Does CSO refrain itself from setting the default SC back to the intree one?

jsafrane replied:
CSO already does not change the default storage class if admin has changed it to something custom. We don't plan any change there.

gnufied commented on enhancements/storage/csi-driver-install.md:
```diff
@@ -30,110 +30,342 @@ superseded-by:
+
+Each CSI driver is deployed via a dedicated operator (i.e. each CSI driver has its own), to accommodate differences
+between the CSI drivers. For example, Manila CSI driver operator can detect if Manila service is present in underlying
+OpenStack cloud and deploy the driver only when the service is present.
```
This says `Manialla CSI driver` can detect if manilla service is present but section below says CSO detects cloudprovider.  So what happens when CSO installs manila operator but the manila operator determines if manila service is not present?

Same thing applies to vsphere CSI driver too.

jsafrane replied:
The Manila operator should advertise in its CR.Status.Conditions that everything is ok (nothing is degraded), but it won't run any Manila CSI driver, because Manila is not there. CSO should translate it into "Available: true", but with a note that only e.g. Cinder CSI driver is running.

jsafrane commented:
Moved the CSI drivers and operators into `openshift-platform-csi-drivers` drivers, still managed by cluster-storage-operator in `openshift-cluster-storage-operator` namespace.

Marked as implementable.

jsafrane commented:
Reworked a lot to cover implementation status in 4.6 - all changes in CSO have been implemented and we have a few operators in 4.6 already (and couple more in development for 4.7).

bertinatto reviewed (commented):
/lgtm
